include/RAJA/policy/cuda/reduce.hpp

```diff
@@ -287,7 +291,54 @@ RAJA_DEVICE RAJA_INLINE void grid_reduce(RAJA::expt::detail::Reducer<OP, T>& red
   const int blockId = BlockIterationGetter::index();
   const int threadId = ThreadIterationGetter::index();
 
-  T temp = block_reduce<ThreadIterationGetter, OP>(red.val, OP::identity());
+  T temp = block_reduce<ThreadIterationGetter, OP>(val, OP<T,T,T>::identity());
+
+  // one thread per block writes to device_mem
+  bool lastBlock = false;
+  if (threadId == 0) {
+    device_mem.set(blockId, temp);
+    // ensure write visible to all threadblocks
+    __threadfence();
+    // increment counter, (wraps back to zero if old count == wrap_around)
+    unsigned int old_count = ::atomicInc(device_count, wrap_around);
+    lastBlock = (old_count == wrap_around);
+  }
+
+  // returns non-zero value if any thread passes in a non-zero value
+  lastBlock = __syncthreads_or(lastBlock);
+
+  // last block accumulates values from device_mem
+  if (lastBlock) {
+    temp = OP<T,T,T>::identity();
+    __threadfence();
+
+    for (int i = threadId; i < numBlocks; i += numThreads) {
+      temp = OP<T,T,T>{}(temp, device_mem.get(i));
+    }
+
+    temp = block_reduce<ThreadIterationGetter, OP>(temp, OP<T,T,T>::identity());
+
+    // one thread returns value
+    if (threadId == 0) {
+      *device_target = temp;
+    }
+  }
+}
+
+template <typename GlobalIterationGetter, template <typename, typename, typename> class OP, typename T>
+RAJA_DEVICE RAJA_INLINE void grid_reduce(RAJA::expt::detail::Reducer<OP, T, T>& red)
```

Comment:
Do we still need this one?

Reply:
Probably not, I'll look in to removing this.